src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,986 @@
+
+	MPI_Barrier(MPI_COMM_WORLD);
+	/* Simulate the conflict with other DTX. */
+	if (arg->myrank == 0)
+		daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC,
+				     DAOS_DTX_RESTART | DAOS_FAIL_ALWAYS,
+				     0, NULL);
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	rc = daos_tx_commit(th, NULL);
+	assert_int_equal(rc, -DER_TX_RESTART);
+
+	/* Not allow new I/O before restart the TX. */
+	req.arg->expect_result = -DER_NO_PERM;
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_restart(th, NULL));
+
+	/* Reset the fail_loc */
+	MPI_Barrier(MPI_COMM_WORLD);
+	if (arg->myrank == 0)
+		daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC, 0,
+				     0, NULL);
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	req.arg->expect_result = 0;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_15(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	print_message("DTX restart because of stale pool map\n");
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+	daos_fail_loc_set(DAOS_DTX_STALE_PM | DAOS_FAIL_ALWAYS);
+	if (arg->myrank == 0)
+		daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC,
+				     DAOS_DTX_STALE_PM | DAOS_FAIL_ALWAYS,
+				     0, NULL);
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	req.arg->expect_result = -DER_TX_RESTART;
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+
+	/* Reset the fail_loc */
+	MPI_Barrier(MPI_COMM_WORLD);
+	daos_fail_loc_set(0);
+	if (arg->myrank == 0)
+		daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC, 0,
+				     0, NULL);
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	/* Not allow new I/O before restart the TX. */
+	req.arg->expect_result = -DER_NO_PERM;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_restart(th, NULL));
+
+	req.arg->expect_result = 0;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_handle_resent(test_arg_t *arg, uint64_t fail_loc)
+{
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	print_message("Resend commit because of lost CPD request\n");
```

Comment:
Sorry, this message is some confuse, should be removed, I will add comments.

The basic test logic is that:

dtx_16: inject fail_loc DAOS_DTX_LOST_RPC_REQUEST on server side to simulate the case that the CPD RPC request lost, then the client has the ability to resend the CPD RPC, and the server side handler will handle it as new RPC since the former 'lost' one is not executed.

dtx_17: inject fail_loc DAOS_DTX_LOST_RPC_REPLY on the server side to simulate the case of the CPD RPC reply lost, then the client has the ability to resend the CPD RPC, and then the server side handler will handle it as resent one since the former one has been executed before.